// @rnmapbox/maps/lib/commonjs/components/MapView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.NATIVE_MODULE_NAME = exports.ANDROID_TEXTURE_NATIVE_MODULE_NAME = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _debounce = require("debounce");
var _utils = require("../utils");
var _filterUtils = require("../utils/filterUtils");
var _Logger = _interopRequireDefault(require("../utils/Logger"));
var _NativeBridgeComponent = _interopRequireDefault(require("./NativeBridgeComponent"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
const MapboxGL = _reactNative.NativeModules.MGLModule;
if (MapboxGL == null) {
  console.error('Native part of Mapbox React Native libraries were not registered properly, double check our native installation guides.');
}
const NATIVE_MODULE_NAME = 'RCTMGLMapView';
exports.NATIVE_MODULE_NAME = NATIVE_MODULE_NAME;
const ANDROID_TEXTURE_NATIVE_MODULE_NAME = 'RCTMGLAndroidTextureMapView';
exports.ANDROID_TEXTURE_NATIVE_MODULE_NAME = ANDROID_TEXTURE_NATIVE_MODULE_NAME;
const styles = _reactNative.StyleSheet.create({
  matchParent: {
    flex: 1
  }
});
const defaultStyleURL = MapboxGL.StyleURL.Street;
/**
 * MapView backed by Mapbox Native GL
 */
class MapView extends (0, _NativeBridgeComponent.default)(_react.default.PureComponent, NATIVE_MODULE_NAME) {
  static defaultProps = {
    projection: 'mercator',
    localizeLabels: false,
    scrollEnabled: true,
    pitchEnabled: true,
    rotateEnabled: true,
    attributionEnabled: true,
    compassEnabled: false,
    compassFadeWhenNorth: false,
    logoEnabled: true,
    scaleBarEnabled: true,
    surfaceView: false,
    regionWillChangeDebounceTime: 10,
    regionDidChangeDebounceTime: 500
  };
  constructor(props) {
    super(props);
    this.logger = _Logger.default.sharedInstance();
    this.logger.start();
    this.state = {
      isReady: null,
      region: null,
      width: 0,
      height: 0,
      isUserInteraction: false
    };
    this._onPress = this._onPress.bind(this);
    this._onLongPress = this._onLongPress.bind(this);
    this._onChange = this._onChange.bind(this);
    this._onLayout = this._onLayout.bind(this);

    // debounced map change methods
    this._onDebouncedRegionWillChange = (0, _debounce.debounce)(this._onRegionWillChange.bind(this), props.regionWillChangeDebounceTime, true);
    this._onDebouncedRegionDidChange = (0, _debounce.debounce)(this._onRegionDidChange.bind(this), props.regionDidChangeDebounceTime);
  }
  componentDidMount() {
    this._setHandledMapChangedEvents(this.props);
  }
  componentWillUnmount() {
    this._onDebouncedRegionWillChange.clear();
    this._onDebouncedRegionDidChange.clear();
    this.logger.stop();
  }
  UNSAFE_componentWillReceiveProps(nextProps) {
    this._setHandledMapChangedEvents(nextProps);
  }
  _setHandledMapChangedEvents(props) {
    if ((0, _utils.isAndroid)() || MapboxGL.MapboxV10) {
      const events = [];
      function addIfHasHandler(name) {
        if (props[`on${name}`] != null) {
          if (MapboxGL.EventTypes[name] == null) {
            if (name === 'DidFailLoadingMap') {
              console.warn(`rnmapbox maps: on${name} is deprecated, please use onMapLoadingError`);
            } else {
              console.warn(`rnmapbox maps: ${name} is not supported`);
            }
          } else {
            events.push(MapboxGL.EventTypes[name]);
            return true;
          }
        }
        return false;
      }
      addIfHasHandler('RegionWillChange');
      addIfHasHandler('RegionIsChanging');
      addIfHasHandler('RegionDidChange');
      addIfHasHandler('UserLocationUpdate');
      addIfHasHandler('WillStartLoadingMap');
      addIfHasHandler('DidFinishLoadingMap');
      addIfHasHandler('MapLoadingError');
      addIfHasHandler('DidFailLoadingMap');
      addIfHasHandler('WillStartRenderingFrame');
      addIfHasHandler('DidFinishRenderingFrame');
      addIfHasHandler('DidFinishRenderingFrameFully');
      addIfHasHandler('WillStartRenderingMap');
      addIfHasHandler('DidFinishRenderingMap');
      addIfHasHandler('DidFinishRenderingMapFully');
      addIfHasHandler('DidFinishLoadingStyle');
      if (addIfHasHandler('MapIdle')) {
        console.warn('onMapIdle is deprecated and will be removed in next beta - please use onRegionDidChange');
        if (props.onRegionDidChange) {
          console.warn('rnmapbox/maps: only one of MapView.onRegionDidChange or onMapIdle is supported');
        }
      }
      if (addIfHasHandler('CameraChanged')) {
        console.warn('onCameraChanged is deprecated and will be removed in next beta - please use onRegionIsChanging');
        if (props.onRegionIsChanging) {
          console.warn('rnmapbox/maps: only one of MapView.onRegionIsChanging or onCameraChanged is supported');
        }
      }
      if (props.onRegionWillChange) {
        console.warn('onRegionWillChange is deprecated and will be removed in v10 - please use onRegionIsChanging');
      }
      this._runNativeCommand('setHandledMapChangedEvents', this._nativeRef, [events]);
    }
  }

  /**
   * Converts a geographic coordinate to a point in the given view’s coordinate system.
   *
   * @example
   * const pointInView = await this._map.getPointInView([-37.817070, 144.949901]);
   *
   * @param {Array<number>} coordinate - A point expressed in the map view's coordinate system.
   * @return {Array}
   */
  async getPointInView(coordinate) {
    const res = await this._runNative('getPointInView', [coordinate]);
    return res.pointInView;
  }

  /**
   * Converts a point in the given view’s coordinate system to a geographic coordinate.
   *
   * @example
   * const coordinate = await this._map.getCoordinateFromView([100, 100]);
   *
   * @param {Array<number>} point - A point expressed in the given view’s coordinate system.
   * @return {Array}
   */
  async getCoordinateFromView(point) {
    const res = await this._runNative('getCoordinateFromView', [point]);
    return res.coordinateFromView;
  }

  /**
   * The coordinate bounds (ne, sw) visible in the user’s viewport.
   *
   * @example
   * const visibleBounds = await this._map.getVisibleBounds();
   *
   * @return {Array}
   */
  async getVisibleBounds() {
    const res = await this._runNative('getVisibleBounds');
    return res.visibleBounds;
  }

  /**
   * Returns an array of rendered map features that intersect with a given point.
   *
   * @example
   * this._map.queryRenderedFeaturesAtPoint([30, 40], ['==', 'type', 'Point'], ['id1', 'id2'])
   *
   * @param  {Array<Number>} coordinate - A point expressed in the map view’s coordinate system.
   * @param  {Array=} filter - A set of strings that correspond to the names of layers defined in the current style. Only the features contained in these layers are included in the returned array.
   * @param  {Array=} layerIDs - A array of layer id's to filter the features by
   * @return {FeatureCollection}
   */

  async queryRenderedFeaturesAtPoint(coordinate) {
    let filter = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
    let layerIDs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
    if (!coordinate || coordinate.length < 2) {
      throw new Error('Must pass in valid coordinate[lng, lat]');
    }
    const res = await this._runNative('queryRenderedFeaturesAtPoint', [coordinate, (0, _filterUtils.getFilter)(filter), layerIDs]);
    if ((0, _utils.isAndroid)()) {
      return JSON.parse(res.data);
    }
    return res.data;
  }

  /**
   * Returns an array of rendered map features that intersect with the given rectangle,
   * restricted to the given style layers and filtered by the given predicate. In v10,
   * passing an empty array will query the entire visible bounds of the map.
   *
   * @example
   * this._map.queryRenderedFeaturesInRect([30, 40, 20, 10], ['==', 'type', 'Point'], ['id1', 'id2'])
   *
   * @param  {Array<Number>} bbox - A rectangle expressed in the map view’s coordinate system. For v10, this can be an empty array to query the visible map area.
   * @param  {Array=} filter - A set of strings that correspond to the names of layers defined in the current style. Only the features contained in these layers are included in the returned array.
   * @param  {Array=} layerIDs -  A array of layer id's to filter the features by
   * @return {FeatureCollection}
   */
  async queryRenderedFeaturesInRect(bbox) {
    let filter = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
    let layerIDs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
    if (bbox != null && (bbox.length === 4 || MapboxGL.MapboxV10 && bbox.length === 0)) {
      const res = await this._runNative('queryRenderedFeaturesInRect', [bbox, (0, _filterUtils.getFilter)(filter), layerIDs]);
      if ((0, _utils.isAndroid)()) {
        return JSON.parse(res.data);
      }
      return res.data;
    } else {
      throw new Error('Must pass in a valid bounding box: [top, right, bottom, left]. An empty array [] is also acceptable in v10.');
    }
  }

  /**
   * Map camera will perform updates based on provided config. Deprecated use Camera#setCamera.
   * @deprecated use Camera#setCamera
   */
  setCamera() {
    console.warn('MapView.setCamera is deprecated - please use Camera#setCamera');
  }
  _runNative(methodName) {
    let args = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
    return super._runNativeCommand(methodName, this._nativeRef, args);
  }

  /**
   * Takes snapshot of map with current tiles and returns a URI to the image
   * @param  {Boolean} writeToDisk If true will create a temp file, otherwise it is in base64
   * @return {String}
   */
  async takeSnap() {
    let writeToDisk = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
    const res = await this._runNative('takeSnap', [writeToDisk]);
    return res.uri;
  }

  /**
   * Returns the current zoom of the map view.
   *
   * @example
   * const zoom = await this._map.getZoom();
   *
   * @return {Number}
   */

  async getZoom() {
    const res = await this._runNative('getZoom');
    return res.zoom;
  }

  /**
   * Returns the map's geographical centerpoint
   *
   * @example
   * const center = await this._map.getCenter();
   *
   * @return {Array<Number>} Coordinates
   */
  async getCenter() {
    const res = await this._runNative('getCenter');
    return res.center;
  }

  /**
   * Queries the currently loaded data for elevation at a geographical location.
   * The elevation is returned in meters relative to mean sea-level.
   * Returns null if terrain is disabled or if terrain data for the location hasn't been loaded yet.
   *
   * @param {Array<Number>} coordinate - the coordinates to query elevation at
   * @return {Number}
   */
  async queryTerrainElevation(coordinate) {
    const res = await this._runNative('queryTerrainElevation', [coordinate]);
    return res.data;
  }

  /**
   * Sets the visibility of all the layers referencing the specified `sourceLayerId` and/or `sourceId`
   *
   * @example
   * await this._map.setSourceVisibility(false, 'composite', 'building')
   *
   * @param {boolean} visible - Visibility of the layers
   * @param {String} sourceId - Identifier of the target source (e.g. 'composite')
   * @param {String=} sourceLayerId - Identifier of the target source-layer (e.g. 'building')
   */
  setSourceVisibility(visible, sourceId) {
    let sourceLayerId = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
    this._runNative('setSourceVisibility', [visible, sourceId, sourceLayerId]);
  }

  /**
   * Show the attribution and telemetry action sheet.
   * If you implement a custom attribution button, you should add this action to the button.
   */
  showAttribution() {
    return this._runNativeCommand('showAttribution', this._nativeRef);
  }
  _onPress(e) {
    if ((0, _utils.isFunction)(this.props.onPress)) {
      this.props.onPress(e.nativeEvent.payload);
    }
  }
  _onLongPress(e) {
    if ((0, _utils.isFunction)(this.props.onLongPress)) {
      this.props.onLongPress(e.nativeEvent.payload);
    }
  }
  _onRegionWillChange(payload) {
    if ((0, _utils.isFunction)(this.props.onRegionWillChange)) {
      this.props.onRegionWillChange(payload);
    }
    this.setState({
      isUserInteraction: payload.properties.isUserInteraction,
      isAnimatingFromUserInteraction: payload.properties.isAnimatingFromUserInteraction
    });
  }
  _onRegionDidChange(payload) {
    if ((0, _utils.isFunction)(this.props.onRegionDidChange)) {
      this.props.onRegionDidChange(payload);
    }
    this.setState({
      region: payload
    });
  }
  _onChange(e) {
    const {
      regionWillChangeDebounceTime,
      regionDidChangeDebounceTime
    } = this.props;
    const {
      type,
      payload
    } = e.nativeEvent;
    let propName = '';
    let deprecatedPropName = '';
    switch (type) {
      case MapboxGL.EventTypes.RegionWillChange:
        if (regionWillChangeDebounceTime && regionWillChangeDebounceTime > 0) {
          this._onDebouncedRegionWillChange(payload);
        } else {
          propName = 'onRegionWillChange';
        }
        break;
      case MapboxGL.EventTypes.RegionIsChanging:
        propName = 'onRegionIsChanging';
        break;
      case MapboxGL.EventTypes.RegionDidChange:
        if (regionDidChangeDebounceTime && regionDidChangeDebounceTime > 0) {
          this._onDebouncedRegionDidChange(payload);
        } else {
          propName = 'onRegionDidChange';
        }
        break;
      case MapboxGL.EventTypes.CameraChanged:
        propName = 'onCameraChanged';
        break;
      case MapboxGL.EventTypes.MapIdle:
        propName = 'onMapIdle';
        break;
      case MapboxGL.EventTypes.UserLocationUpdated:
        propName = 'onUserLocationUpdate';
        break;
      case MapboxGL.EventTypes.WillStartLoadingMap:
        propName = 'onWillStartLoadingMap';
        break;
      case MapboxGL.EventTypes.DidFinishLoadingMap:
        propName = 'onDidFinishLoadingMap';
        break;
      case MapboxGL.EventTypes.DidFailLoadingMap:
        propName = 'onDidFailLoadingMap';
        break;
      case MapboxGL.EventTypes.MapLoadingError:
        propName = 'onMapLoadingError';
        deprecatedPropName = 'onDidFailLoadingMap';
        break;
      case MapboxGL.EventTypes.WillStartRenderingFrame:
        propName = 'onWillStartRenderingFrame';
        break;
      case MapboxGL.EventTypes.DidFinishRenderingFrame:
        propName = 'onDidFinishRenderingFrame';
        break;
      case MapboxGL.EventTypes.DidFinishRenderingFrameFully:
        propName = 'onDidFinishRenderingFrameFully';
        break;
      case MapboxGL.EventTypes.WillStartRenderingMap:
        propName = 'onWillStartRenderingMap';
        break;
      case MapboxGL.EventTypes.DidFinishRenderingMap:
        propName = 'onDidFinishRenderingMap';
        break;
      case MapboxGL.EventTypes.DidFinishRenderingMapFully:
        propName = 'onDidFinishRenderingMapFully';
        break;
      case MapboxGL.EventTypes.DidFinishLoadingStyle:
        propName = 'onDidFinishLoadingStyle';
        break;
      default:
        console.warn('Unhandled event callback type', type);
    }
    if (propName !== '') {
      this._handleOnChange(propName, payload);
    }
    if (deprecatedPropName !== '') {
      this._handleOnChange(deprecatedPropName, payload);
    }
  }
  _onLayout(e) {
    this.setState({
      isReady: true,
      width: e.nativeEvent.layout.width,
      height: e.nativeEvent.layout.height
    });
  }
  _handleOnChange(propName, payload) {
    const func = this.props[propName];
    if (func && (0, _utils.isFunction)(func)) {
      func(payload);
    }
  }
  _getContentInset() {
    if (!this.props.contentInset) {
      return;
    }
    if (!Array.isArray(this.props.contentInset)) {
      return [this.props.contentInset];
    }
    return this.props.contentInset;
  }
  _setNativeRef(nativeRef) {
    this._nativeRef = nativeRef;
    super._runPendingNativeCommands(nativeRef);
  }
  setNativeProps(props) {
    if (this._nativeRef) {
      this._nativeRef.setNativeProps(props);
    }
  }
  _setStyleURL(props) {
    // user set a styleURL, no need to alter props
    if (props.styleURL) {
      return;
    }

    // user set styleJSON pass it to styleURL
    if (props.styleJSON && !props.styleURL) {
      props.styleURL = props.styleJSON;
    }

    // user neither set styleJSON nor styleURL
    // set defaultStyleUrl
    if (!props.styleJSON || !props.styleURL) {
      props.styleURL = defaultStyleURL;
    }
  }
  render() {
    const props = {
      ...this.props,
      contentInset: this._getContentInset(),
      style: styles.matchParent
    };
    this._setStyleURL(props);
    const callbacks = {
      ref: nativeRef => this._setNativeRef(nativeRef),
      onPress: this._onPress,
      onLongPress: this._onLongPress,
      onMapChange: this._onChange,
      onAndroidCallback: (0, _utils.isAndroid)() ? this._onAndroidCallback : undefined
    };
    let mapView = null;
    if ((0, _utils.isAndroid)() && !this.props.surfaceView && this.state.isReady) {
      mapView = /*#__PURE__*/_react.default.createElement(RCTMGLAndroidTextureMapView, _extends({}, props, callbacks), this.props.children);
    } else if (this.state.isReady) {
      mapView = /*#__PURE__*/_react.default.createElement(RCTMGLMapView, _extends({}, props, callbacks), this.props.children);
    }
    return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
      onLayout: this._onLayout,
      style: this.props.style,
      testID: mapView ? undefined : this.props.testID
    }, mapView);
  }
}
const RCTMGLMapView = (0, _reactNative.requireNativeComponent)(NATIVE_MODULE_NAME);
let RCTMGLAndroidTextureMapView;
if ((0, _utils.isAndroid)()) {
  RCTMGLAndroidTextureMapView = (0, _reactNative.requireNativeComponent)(ANDROID_TEXTURE_NATIVE_MODULE_NAME);
}
var _default = MapView;
exports.default = _default;
