// @rnmapbox/maps/lib/commonjs/components/annotations/Annotation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _propTypes = _interopRequireDefault(require("prop-types"));
var _SymbolLayer = require("../SymbolLayer");
var _Animated = _interopRequireDefault(require("../../utils/animated/Animated"));
var _classes = require("../../classes");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
class Annotation extends _react.default.Component {
  static propTypes = {
    id: _propTypes.default.string.isRequired,
    animated: _propTypes.default.bool,
    animationDuration: _propTypes.default.number,
    animationEasingFunction: _propTypes.default.func,
    coordinates: _propTypes.default.arrayOf(_propTypes.default.number),
    onPress: _propTypes.default.func,
    children: _propTypes.default.any,
    style: _propTypes.default.any,
    icon: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number, _propTypes.default.object])
  };
  static defaultProps = {
    animated: false,
    animationDuration: 1000,
    animationEasingFunction: _reactNative.Easing.linear
  };
  constructor(props) {
    super(props);
    const shape = this._getShapeFromProps(props);
    this.state = {
      shape: props.animated ? new _classes.AnimatedPoint(shape) : shape
    };
    this.onPress = this.onPress.bind(this);
  }
  componentDidUpdate(prevProps) {
    if (!Array.isArray(this.props.coordinates)) {
      this.setState({
        shape: null
      });
      return;
    }
    const hasCoordChanged = prevProps.coordinates[0] !== this.props.coordinates[0] || prevProps.coordinates[1] !== this.props.coordinates[1];
    if (prevProps.animated !== this.props.animated || hasCoordChanged && (!this.state.shape || !this.props.animated)) {
      const shape = this._getShapeFromProps(this.props);
      this.setState({
        shape: this.props.animated ? new _classes.AnimatedPoint(shape) : shape
      });
    } else if (hasCoordChanged && this.props.animated && this.state.shape) {
      // flush current animations
      this.state.shape.stopAnimation();
      this.state.shape.timing({
        coordinates: this.props.coordinates,
        easing: this.props.animationEasingFunction,
        duration: this.props.animationDuration
      }).start();
    }
  }
  onPress() {
    if (this.props.onPress) {
      this.props.onPress();
    }
  }
  _getShapeFromProps() {
    let props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    const lng = props.coordinates[0] || 0;
    const lat = props.coordinates[1] || 0;
    return {
      type: 'Point',
      coordinates: [lng, lat]
    };
  }
  get symbolStyle() {
    if (!this.props.icon) {
      return undefined;
    }
    return Object.assign({}, this.props.style, {
      iconImage: this.props.icon
    });
  }
  render() {
    if (!this.props.coordinates) {
      return null;
    }
    return /*#__PURE__*/_react.default.createElement(_Animated.default.ShapeSource, {
      id: this.props.id,
      ref: "source",
      onPress: this.onPress,
      shape: this.state.shape
    }, this.symbolStyle && /*#__PURE__*/_react.default.createElement(_SymbolLayer.SymbolLayer, {
      id: `${this.props.id}-symbol`,
      style: this.symbolStyle
    }), this.props.children);
  }
}
var _default = Annotation;
exports.default = _default;
