// @rnmapbox/maps/lib/commonjs/utils/geoUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.addToFeatureCollection = addToFeatureCollection;
exports.calculateDistance = void 0;
exports.getOrCalculateVisibleRegion = getOrCalculateVisibleRegion;
exports.makeFeature = void 0;
exports.makeFeatureCollection = makeFeatureCollection;
exports.makeLatLngBounds = makeLatLngBounds;
exports.pointAlongLine = exports.makePoint = exports.makeLineString = void 0;
var _helpers = require("@turf/helpers");
var _distance = _interopRequireDefault(require("@turf/distance"));
var _along = _interopRequireDefault(require("@turf/along"));
var _geoViewport = _interopRequireDefault(require("@mapbox/geo-viewport"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const VECTOR_TILE_SIZE = 512;
const makePoint = _helpers.point;
exports.makePoint = makePoint;
const makeLineString = _helpers.lineString;
exports.makeLineString = makeLineString;
function makeLatLngBounds(northEastCoordinates, southWestCoordinates) {
  return (0, _helpers.featureCollection)([(0, _helpers.point)(northEastCoordinates), (0, _helpers.point)(southWestCoordinates)]);
}
const makeFeature = _helpers.feature;
exports.makeFeature = makeFeature;
function makeFeatureCollection() {
  let features = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
  let options = arguments.length > 1 ? arguments[1] : undefined;
  return (0, _helpers.featureCollection)(features, options);
}
function addToFeatureCollection(newFeatureCollection, newFeature) {
  return {
    ...newFeatureCollection,
    features: [...newFeatureCollection.features, newFeature]
  };
}
const calculateDistance = _distance.default;
exports.calculateDistance = calculateDistance;
const pointAlongLine = _along.default;
exports.pointAlongLine = pointAlongLine;
function getOrCalculateVisibleRegion(coord, zoomLevel, width, height, nativeRegion) {
  const region = {
    ne: [0, 0],
    sw: [0, 0]
  };
  if (!nativeRegion || !Array.isArray(nativeRegion.properties.visibleBounds)) {
    const bounds = _geoViewport.default.bounds(coord, zoomLevel, [width, height], VECTOR_TILE_SIZE);
    region.ne = [bounds[3], bounds[2]];
    region.sw = [bounds[1], bounds[0]];
  } else {
    [region.ne, region.sw] = nativeRegion.properties.visibleBounds;
  }
  return region;
}
